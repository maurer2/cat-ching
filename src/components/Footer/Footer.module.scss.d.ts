import globalClassNames, { ClassNames as GlobalClassNames } from '../..style.d';
declare const classNames: typeof globalClassNames & {
  readonly footer: 'footer';
  readonly detail: 'detail';
  readonly detailKey: 'detailKey';
  readonly detailValue: 'detailValue';
  readonly toggleButton: 'toggleButton';
};
export default classNames;
export type ClassNames =
  | 'footer'
  | 'detail'
  | 'detailKey'
  | 'detailValue'
  | 'toggleButton'
  | GlobalClassNames;
