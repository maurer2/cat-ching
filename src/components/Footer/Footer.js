import React, { useState } from 'react';
import PropTypes from 'prop-types';

import style from './Footer.module.scss';

function Footer({ currentAmount }) {
  const [showCurrentAmount, setShowCurrentAmount] = useState(true);

  function toggleAmountVisbility() {
    setShowCurrentAmount(!showCurrentAmount);
  }

  return (
    <footer className={style.footer}>
      {showCurrentAmount && (
        <dl className={style.detail}>
          <dt className={style.detailKey}>
            Current amount:
          </dt>
          <dd className={style.detailValue}>
            {currentAmount.valueFormatted}
          </dd>
        </dl>
      )}
      <button
        className={style.toggleButton}
        onClick={toggleAmountVisbility}
        type="button"
      >
        {showCurrentAmount ? 'Hide current amount' : 'Show current amount' }
      </button>
    </footer>
  );
}

export default Footer;

const { shape, number, string } = PropTypes;

Footer.propTypes = {
  currentAmount: shape({
    value: number.isRequired,
    currency: string.isRequired,
    valueInCents: number.isRequired,
    valueFormatted: string.isRequired,
  }).isRequired,
};
