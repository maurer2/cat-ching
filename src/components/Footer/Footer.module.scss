.footer {
  display: flex;
  padding: 1rem;
  align-items: center;
  background: black;
  color: white;
}

.detail {
  display: flex;
  margin: 0;
  flex-wrap: wrap;
  flex-grow: 1;

  & + .detail {
    margin-top: 0.5rem;
  }
}

.detailKey {
  margin-right: 0.5rem;
  flex-basis: auto;
}

.detailValue {
  margin: 0;
  font-weight: bold;
}

.toggleButton {
  all: unset;
  display: flex;
  flex-direction: column;
  margin-left: auto;
  padding: 0.5rem;
  background: white;
  color: black;
  border: 1px solid currentColor;
  border-radius: 0;
  text-align: center;
  cursor: pointer;

  &:hover,
  &:active {
    border: 1px solid black;
  }

  &:active {
    border: 1px solid white;
  }

  > .hidden-text {
    overflow: hidden;
    height: 0;
  }
}