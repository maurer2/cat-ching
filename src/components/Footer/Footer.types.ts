import type Money from '../../types/Money';

export type FooterProps = {
  currentAmount: Money,
};
