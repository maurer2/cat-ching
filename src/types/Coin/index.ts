export type { CoinData, Coin } from './Coin';
