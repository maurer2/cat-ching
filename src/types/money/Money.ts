/* eslint-disable no-underscore-dangle */
/* eslint-disable @typescript-eslint/lines-between-class-members */
export const currencies = {
  GBP: 'Â£',
  USD: '$',
} as const;

export type CurrencyName = keyof typeof currencies;

export default class Money {
  private readonly _value: bigint;
  private readonly _name: CurrencyName;
  private readonly _formatter: Intl.NumberFormat;

  public static fromBigInt(
    valueInFractions: bigint,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    return new this(valueInFractions, name, formatter);
  }

  public static fromNumber(
    valueInFractions: number,
    name: CurrencyName,
    formatter?: Intl.NumberFormat,
  ): Money {
    const valueAsBigInt = BigInt(valueInFractions);

    return new this(valueAsBigInt, name, formatter);
  }

  public static fromRandom(name: CurrencyName, formatter?: Intl.NumberFormat): Money {
    const randomNumber: number = Math.ceil(Math.random() * (1000 - 1)) + 1;
    const valueAsBigInt = BigInt(randomNumber);

    return new this(valueAsBigInt, name, formatter);
  }

  private constructor(valueInFractions: bigint, name: CurrencyName, formatter?: Intl.NumberFormat) {
    this._value = valueInFractions;
    this._name = name;
    this._formatter = formatter
      ?? new Intl.NumberFormat('en-GB', {
        style: 'currency',
        currency: 'GBP',
        useGrouping: false, // disable thousand separator
      });
  }

  public add(valueToAdd: Money): Money {
    const newValue: bigint = this._value + valueToAdd._value;

    return new Money(newValue, this._name, this._formatter);
  }

  public isEqualTo(secondValue: Money): boolean {
    return this._value === secondValue._value;
  }

  public isNegative(): boolean {
    return this._value < 0;
  }

  public get valueAsSeparateParts(): Intl.NumberFormatPart[] {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.formatToParts(valueAsNumber);
  }

  public get valueAsFormattedString(): string {
    // todo Number.MAX_SAFE_INTEGER and Number.MIN_SAFE_INTEGER check
    const valueAsNumber = Number(this._value) / 100;

    return this._formatter.format(valueAsNumber);
  }

  public get valueFormattedAsIntegerAndFraction(): string {
    // https://stackoverflow.com/a/70086020
    const integer = Number(this._value / 100n);
    const fraction = Number(this._value % 100n);
    // const integral = Number(this.value / 100n);
    // const fraction = Number(this.value % 100n);

    return `${integer}.${`${fraction}`.padEnd(2, '0')}`;
  }
}

export type MoneyType = typeof Money;
