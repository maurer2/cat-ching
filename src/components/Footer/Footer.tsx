import React, { useReducer } from 'react';

import style from './Footer.module.scss';
import * as Types from './Footer.types';

function Footer({ currentAmount }: Types.FooterProps): JSX.Element {
  const [showCurrentAmount, toggleShowCurrentAmount] = useReducer((state: boolean) => !state, true);

  return (
    <footer className={style.footer}>
      {showCurrentAmount && (
        <dl className={style.detail}>
          <dt className={style.detailKey}>Current amount:</dt>
          <dd className={style.detailValue}>{currentAmount.valueAsFormattedString}</dd>
        </dl>
      )}
      <button
        className={style.toggleButton}
        onClick={toggleShowCurrentAmount}
        type="button"
      >
        {showCurrentAmount ? 'Hide current amount' : 'Show current amount'}
      </button>
    </footer>
  );
}

export default Footer;
