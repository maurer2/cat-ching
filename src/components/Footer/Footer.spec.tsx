/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Component from './Footer';
import * as Types from './Footer.types';

import Money from '../../types/Money';

describe('AppContainer', () => {
  const propsDefault: Types.FooterProps = {
    currentAmount: Money.fromNumber(111, 'GBP'),
  };

  const setup = (props = {}) => render(<Component {...propsDefault} {...props} />);

  it('renders', () => {
    const { queryByTestId } = setup();

    expect(queryByTestId('footer')).toBeInTheDocument();
  });

  it('has child elements', () => {
    const { getByRole } = setup();

    expect(getByRole('button')).toBeInTheDocument();
  });

  it('amount is is visible on load', () => {
    const { getByText } = setup();

    expect(getByText('Current amount:')).toBeInTheDocument();
    expect(getByText('£1.11')).toBeInTheDocument();
  });

  it('triggers amount visibility on click', async () => {
    const user = userEvent.setup();
    const { getByRole, queryByText } = setup();

    await user.click(getByRole('button'));
    expect(queryByText('Current amount:')).not.toBeInTheDocument();
    expect(queryByText('£1.11')).not.toBeInTheDocument();
  });
});
