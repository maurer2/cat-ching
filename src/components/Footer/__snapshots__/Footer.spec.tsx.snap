// Vitest Snapshot v1

exports[`AppContainer > matches snapshot 1`] = `
<div>
  <footer
    class="_footer_1j6yp_1"
    data-testid="footer"
  >
    <dl
      class="_detail_1j6yp_9"
    >
      <dt
        class="_detailKey_1j6yp_19"
      >
        Current amount:
      </dt>
      <dd
        class="_detailValue_1j6yp_24"
      >
        Â£1.11
      </dd>
    </dl>
    <button
      class="_toggleButton_1j6yp_29"
      type="button"
    >
      Hide current amount
    </button>
  </footer>
</div>
`;
